File: macos/Sources/xml-macker/DocumentSession.swift
import Cocoa

// One open file = one session (Chrome-style tabs, v0.35.0).
//
// Ownership contract: the ACTIVE tab's canonical state lives in the
// live UI (SourceViewController's text storage, MainWindowController's
// currentTree/currentSelectedNode). A session's fields below are the
// PARKED snapshot, captured by snapshotActiveSession() at the moment
// the user switches away, and poured back by switchToTab(). So a
// freshly created session is mostly empty until its first switch-away.
//
// Memory: a parked session keeps its full NSTextStorage and parsed
// tree alive so switching back is instant (no re-read, no re-parse).
// That's the Chrome trade: N open tabs cost N documents of RAM.
final class DocumentSession {
    // Stable identity keeps asynchronous load/save/validation completions
    // attached to the tab that started them, even after tab switches or Save As.
    let id = UUID()
    var url: URL                    // var: Save As… re-points the active tab
    /// A document that has never been saved anywhere the user chose. It
    /// lives as a real file in a private scratch folder so the ordinary
    /// file-loading path works unchanged, but Save asks where to put it,
    /// it never joins Recent Files or the reopen-at-launch list, and the
    /// scratch copy is thrown away when the app quits.
    var isUntitled = false
    /// The marker strokes of this document, kept per tab so switching
    /// away and back brings them with it.
    let highlights = TextHighlights()
    var fileSize: Int
    var textEncoding: XMLTextEncoding = .utf8
    var fileModificationDate: Date?

    var storage: NSTextStorage?
    var lineStarts: [Int] = [0]
    var tree: XMLTreeNode?
    var parseErrors: [XMLStreamParser.ParseError] = []
    var selectedNode: XMLTreeNode?
    var scrollOrigin: NSPoint = .zero
    var isDirty = false
    // Incremented for every edit. A save clears dirty state only when this
    // revision still matches the snapshot that was actually written.
    var editRevision: UInt64 = 0
    // Set when the document was edited while PARKED (diff copy-hunk):
    // the snapshot tree is stale, so activation triggers a reparse.
    var needsReparse = false

    // Each tab has its OWN undo stack, a shared one would try to
    // replay edit ranges from one document inside another.
    let undoManager = UndoManager()

    // True from loadFile until the parse + text load complete; tab
    // switching is refused while any load is in flight.
    var isLoading = false

    init(url: URL, fileSize: Int) {
        self.url = url
        self.fileSize = fileSize
    }
}

File: windows/src/xml-macker/Core/DocumentSession.cs
using System;
using System.Collections.Generic;
using XMLMacker.Editor;

namespace XMLMacker.Core;

/// <summary>
/// Per-tab parked state (Chrome-style tabs, one open file = one session = one tab). A 1:1 port of the Swift
/// <c>DocumentSession</c>.
///
/// The ACTIVE tab's canonical state lives in the live UI (the editor's text buffer and the window controller's
/// current tree/selection); a session's fields are filled by the "snapshot on switch-away" step and read back on
/// "switch-to". A freshly created session is mostly empty until its first switch-away.
///
/// Memory trade ("the Chrome trade"): a parked session keeps its full text buffer AND parsed tree alive so
/// switching back is instant (no re-read, no re-parse). N open tabs cost N documents of RAM.
///
/// Behavioral contracts the owner (session/tab manager) must honor:
/// <list type="bullet">
/// <item>Snapshot on switch-away, restore on switch-to (the session isn't canonical until snapshotted).</item>
/// <item>Refuse tab switches while any session has <see cref="IsLoading"/> == <c>true</c>.</item>
/// <item>On activation, if <see cref="NeedsReparse"/>, reparse before showing the tree.</item>
/// <item><see cref="LineStarts"/> must be maintained incrementally on edit (a full 12.6M-line rescan per keystroke is not viable).</item>
/// </list>
/// </summary>
public sealed class DocumentSession
{
    public Guid Id { get; } = Guid.NewGuid();

    /// <summary>File path. <b>Mutable</b>, "Save As…" re-points the active tab.</summary>
    public string Url { get; set; }

    /// <summary>File size in bytes.</summary>
    public long FileSize { get; set; }

    public XmlTextEncoding TextEncoding { get; set; } = XmlTextEncoding.Utf8;

    public DateTime? FileModificationDateUtc { get; set; }

    /// <summary>
    /// The tab's full editable text buffer (parked). The WPF document text model, see <see cref="ITextBuffer"/>.
    /// <c>null</c> until the tab has content.
    /// </summary>
    public ITextBuffer? Storage { get; set; }

    /// <summary>
    /// Line-offset table: <c>LineStarts[k]</c> = char offset where line <c>k+1</c> begins. Enables O(1)
    /// line↔offset mapping for a 12.6M-line file without scanning. Always starts <c>[0]</c> (line 1 begins at offset 0).
    /// </summary>
    public int[] LineStarts { get; set; } = new[] { 0 };

    /// <summary>Parsed tree root (parked). <c>null</c> until parsed.</summary>
    public XmlTreeNode? Tree { get; set; }

    /// <summary>Last full-parse errors.</summary>
    public IReadOnlyList<ParseError> ParseErrors { get; set; } = Array.Empty<ParseError>();

    /// <summary>Last selected tree node.</summary>
    public XmlTreeNode? SelectedNode { get; set; }

    /// <summary>Editor scroll position to restore (the WPF analog of <c>NSPoint</c>). Defaults to <c>(0, 0)</c>.</summary>
    public (double X, double Y) ScrollOrigin { get; set; }

    /// <summary>Unsaved-changes flag.</summary>
    public bool IsDirty { get; set; }

    public ulong EditRevision { get; set; }

    /// <summary>The marker strokes (the highlighter); the main window stores them per file.</summary>
    public TextHighlights Highlights { get; } = new();

    /// <summary>
    /// Set when the doc was edited while PARKED (e.g. a diff copy-hunk applied to a background tab): the snapshot
    /// tree is stale, so activation must trigger a reparse.
    /// </summary>
    public bool NeedsReparse { get; set; }

    /// <summary>
    /// This tab's OWN undo stack. A shared stack would replay one document's edit ranges against another, so each
    /// tab owns its own instance. Typed as <see cref="object"/> because the concrete <c>EditorUndoStack</c> lives in
    /// the editor module; the session/editor assigns it.
    /// </summary>
    public object? UndoStack { get; set; }

    /// <summary>True from load start until parse + text load complete. Tab switching is refused while this is set for any session.</summary>
    public bool IsLoading { get; set; }

    /// <summary>Creates a session for <paramref name="url"/>; all other fields keep their defaults.</summary>
    public DocumentSession(string url, long fileSize)
    {
        Url = url;
        FileSize = fileSize;
    }
}
